File: examples/python/testing/examples/sequential-workflow.py
from aws_durable_execution_sdk_python import DurableContext, durable_execution, durable_step
from aws_durable_execution_sdk_python.types import StepContext
from aws_durable_execution_sdk_python.execution import InvocationStatus
from aws_durable_execution_sdk_python.lambda_service import OperationType
from aws_durable_execution_sdk_python_testing.runner import DurableFunctionTestRunner


@durable_step
def validate(ctx: StepContext, order_id: str) -> dict:
    return {"order_id": order_id, "status": "validated"}


@durable_step
def payment(ctx: StepContext, order: dict) -> dict:
    return {**order, "payment": "completed"}


@durable_step
def fulfillment(ctx: StepContext, order: dict) -> dict:
    return {**order, "fulfillment": "shipped"}


@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    validated = context.step(validate(event["order_id"]), name="validate")
    paid = context.step(payment(validated), name="payment")
    return context.step(fulfillment(paid), name="fulfillment")


def test_executes_all_steps_in_order():
    runner = DurableFunctionTestRunner(handler=handler)
    with runner:
        result = runner.run(input='{"order_id": "order-123"}', timeout=10)

    assert result.status is InvocationStatus.SUCCEEDED

    step_ops = [op for op in result.operations if op.operation_type == OperationType.STEP]
    assert len(step_ops) == 3
    assert [op.name for op in step_ops] == ["validate", "payment", "fulfillment"]

File: examples/typescript/testing/examples/sequential-workflow.ts
import { withDurableExecution, DurableContext } from "@aws/durable-execution-sdk-js";
import { LocalDurableTestRunner } from "@aws/durable-execution-sdk-js-testing";
import { ExecutionStatus, OperationType } from "@aws-sdk/client-lambda";

const handler = withDurableExecution(async (event: { orderId: string }, context: DurableContext) => {
  const validated = await context.step("validate", () => ({ orderId: event.orderId, status: "validated" }));
  const paid = await context.step("payment", () => ({ ...validated, payment: "completed" }));
  const fulfilled = await context.step("fulfillment", () => ({ ...paid, fulfillment: "shipped" }));
  return fulfilled;
});

let runner: LocalDurableTestRunner;

beforeAll(async () => {
  await LocalDurableTestRunner.setupTestEnvironment();
});

afterAll(async () => {
  await LocalDurableTestRunner.teardownTestEnvironment();
});

beforeEach(() => {
  runner = new LocalDurableTestRunner({ handlerFunction: handler });
});

it("executes all steps in order", async () => {
  const result = await runner.run({ payload: { orderId: "order-123" } });

  expect(result.getStatus()).toBe(ExecutionStatus.SUCCEEDED);

  const ops = result.getOperations().filter(op => op.getType() === OperationType.STEP);
  expect(ops.length).toBe(3);
  expect(ops.map(op => op.getName())).toEqual(["validate", "payment", "fulfillment"]);
});
